core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -110,8 +124,24 @@ object Semaphore {
         def withPermitsScoped(n: Long)(implicit trace: Trace): ZIO[Scope, Nothing, Unit] =
           ZIO.acquireRelease(reserve(n))(_.release).flatMap(_.acquire)
 
+        override def tryWithPermits[R, E, A](n: Long)(zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, Option[A]] =
+          tryReserve(n).flatMap {
+            case Some(reservation) => (reservation.acquire *> zio <* reservation.release).asSome
+            case None              => ZIO.succeed(None)
```

Comment:
more performant this way

```suggestion
            case _                 => Exit.none
```